### ISSUE TITLE
Returning undefined triggers no-return-assign

### ISSUE THREAD

**Issue by slikts** [eslint/eslint#10824]
* **ESLint Version:** 5.5.0
* **Node Version:** 10.6.0
* **npm Version:** 6.4.1

Default configuration; `() => void (a = 1)` triggers the no-return-assign rule even though it's returning `undefined`.

**Comment by ljharb** [comment]
the rule warns on returning assignments; those can return `undefined` too. Using void doesn’t make it “not returning an assignment”.

**Comment by slikts** [comment]
It looks like this:
```
> npx eslint --version
v5.5.0
> npx eslint .

example.js
  error  Arrow function should not return assignment  no-return-assign
```
And example.js is: 
```javascript
const foo = () => void (global.a = 2)
foo()
```
I've not been able to replicate this in the demo.

**Comment by ljharb** [comment]
The rule isn't warning about returning the *result* of an assignment - it's warning about having an assignment expression inside a return statement. Adding `void` doesn't change that.

**Comment by ilyavolodin** [comment]
I agree with @ljharb this seems to be intentional behavior of the rule. Because you are using short form of the arrow your code can be re-written as:
```
() => { 
  return void a = 1;
}
```
or even to:
```
function () {
  return void a = 1;
}
```

Both of those should be flagged since you are doing the assignment in the return statement.
